### PR TITLE
[SDP-1148] Add `base_url` and `sdp_ui_base_url` default values

### PR DESCRIPTION
**Pull request by ziyliu**
### What
- Make `base_url` and `sdp_ui_base_url` optional fields in the `POST /tenants` request body. Skip validation on fields if left blank
- If left blank in the request, handler when dynamically generate the url's based on the name in the request and the base url attached to the server.

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### Known limitations

[TODO or N/A]

#### PR Structure

* [x] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR title and description are clear enough for anyone to review it.
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [x] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: stellar-multitenant/internal/httphandler/tenants_handler.go [hunk 1]
```diff
@@ -97,9 +98,28 @@ func (h TenantsHandler) Post(rw http.ResponseWriter, req *http.Request) {
...
+			httperror.InternalError(ctx, "Could not generate tenant URL", err, nil).Render(rw)
+			return
+		}
+	}
+
+	var tntSDPUIBaseURL string
+	if reqBody.SDPUIBaseURL != nil {
+		tntSDPUIBaseURL = *reqBody.SDPUIBaseURL
+	} else {
+		tntSDPUIBaseURL = tntBaseURL
```
Comment (marwen-abid) [inline stellar-multitenant/internal/httphandler/tenants_handler.go:116]: SDPUIBaseURL should generated the same way as BaseURL above.
tntSDPUIBaseURL, err = utils.GenerateTenantURL(h.SDPBUIBaseURL, tnt.Name)

Comment (ziyliu) [reply stellar-multitenant/internal/httphandler/tenants_handler.go]: thanks for the catch! I addressed it here [708a4a3](https://github.com/stellar/stellar-disbursement-platform-backend/pull/281/commits/708a4a36a972b2ec2c325c6901a224a5dc01bf0f)

Comment (ziyliu) [reply stellar-multitenant/internal/httphandler/tenants_handler.go]: added here! [d2b153d](https://github.com/stellar/stellar-disbursement-platform-backend/pull/281/commits/d2b153d7d3fda0f1dd53857e454491ccc0c201d4)

File: internal/utils/url.go [hunk 2]
```diff
@@ -89,3 +91,13 @@ func GetURLWithScheme(rawURL string) (string, error) {
...
+
+func GenerateTenantURL(baseURL string, tenantID string) (string, error) {
+	protocol, urlName, validURL := strings.Cut(baseURL, urlSeparator)
+	if !validURL || protocol == "" || urlName == "" {
+		return "", fmt.Errorf("invalid base URL: %s", baseURL)
+	}
+
+	tntURL := protocol + urlSeparator + tenantID + "." + urlName
+	return tntURL, nil
+}
```
Comment (marwen-abid) [inline internal/utils/url.go:103]: I think this would be more robust if we rely on `url.Parse()` for parsing and validating the URL. There are a lot of edge cases we'd have to consider otherwise. This should pass the unit tests. 

```go
func GenerateTenantURL(baseURL string, tenantID string) (string, error) {
	if tenantID == "" {
		return "", fmt.Errorf("tenantID is empty")
	}

	parsedURL, err := url.Parse(baseURL)
	if err != nil {
		return "", fmt.Errorf("invalid base URL %s: %w", baseURL, err)
	}

	hostParts := strings.SplitN(parsedURL.Hostname(), ".", 2)
	if len(hostParts) < 2 {
		return "", fmt.Errorf("base URL must have at least two domain parts %s", baseURL)
	}

	newHostname := fmt.Sprintf("%s.%s.%s", tenantID, hostParts[0], hostParts[1])
	if parsedURL.Port() != "" {
		parsedURL.Host = newHostname + ":" + parsedURL.Port()
	} else {
		parsedURL.Host = newHostname
	}

	return parsedURL.String(), nil
}
```

File: internal/utils/url_test.go [hunk 3]
```diff
@@ -180,3 +180,59 @@ func Test_GetURLWithScheme(t *testing.T) {
...
+			expectedErr: fmt.Errorf("invalid base URL: bluecorp.org.local:3000"),
+		},
+		{
+			name:        "returns error when invalid base URL - no protocol",
+			baseURL:     "://bluecorp.org.local:3000",
+			tenantID:    tenantID,
+			expectedURL: "",
+			expectedErr: fmt.Errorf("invalid base URL: ://bluecorp.org.local:3000"),
+		},
+	}
```
Comment (marwen-abid) [inline internal/utils/url_test.go:225]: Nice tests! Maybe adding one with `https` to show that different protocols are handled.

**Review by marwen-abid — APPROVED** [review]
LGTM! Great tests. Just added a couple of comments.

File: internal/utils/url.go [hunk 4]
```diff
@@ -89,3 +91,13 @@ func GetURLWithScheme(rawURL string) (string, error) {
 
 	return rawURL, nil
 }
+
+func GenerateTenantURL(baseURL string, tenantID string) (string, error) {
+	protocol, urlName, validURL := strings.Cut(baseURL, urlSeparator)
+	if !validURL || protocol == "" || urlName == "" {
+		return "", fmt.Errorf("invalid base URL: %s", baseURL)
```
Comment (marwen-abid) [inline internal/utils/url.go:98]: @marcelosalloum brought my attention before to the fact that `:` is used in golang for error wrapping. So it's best to avoid it and only use it in that case. I.e. `"this value is invalid: %w", err` 

https://github.com/stellar/stellar-disbursement-platform-backend/pull/230/files#r1562553009

Comment (ziyliu) [reply internal/utils/url.go:98]: good point, I'll remove the colon

File: cmd/serve_test.go [hunk 5]
```diff
@@ -215,6 +215,8 @@ func Test_serve(t *testing.T) {
 		TenantAccountNativeAssetBootstrapAmount: tenant.MinTenantDistributionAccountAmount,
 		AdminAccount:                            "admin-account",
 		AdminApiKey:                             "admin-api-key",
+		BaseURL:                                 "https://sdp.com",
+		SDPUIBaseURL:                            "http://sdp-ui.com",
```
[the remark this answers is not in the document]
Comment (ziyliu) [reply cmd/serve_test.go]: fixed here [57d266b](https://github.com/stellar/stellar-disbursement-platform-backend/pull/281/commits/57d266b438031a79905ab5c8311aac6d8cc907e4)

File: internal/utils/url.go (hunk 2 above)
Comment (ziyliu) [reply internal/utils/url.go:116]: thanks, these are great suggestions for checking the URL's integrity [8670fc1](https://github.com/stellar/stellar-disbursement-platform-backend/pull/281/commits/8670fc15de3949ca544acde78864904bec6d22e7)
